### PR TITLE
added alternate method to direct reply using the email subject

### PR DESCRIPTION
**Pull request by ruKurz**
This PR is motivated by adding an alternate method in order to direct reply to offline messages via email using the email subject instead of the email to.

**Current situation**
The Rocket.Chat message (ID) to reply to is currently referenced using email subaddressing ([RFC 5233](https://tools.ietf.org/html/rfc5233)). Unfortunately there are a lot of email service providers out there not supporting subaddressing (e.g. [Microsoft O365](https://office365.uservoice.com/forums/264636-general/suggestions/18185419-add-support-for-rfc-5233-subaddressing-myemail-an)). Therefore it might be a good idea to use the `subject` email header in order to pass the message id.

**Changes**
This PR makes it configureable to choose whether to use subaddressing or the email subject within your email reply in order to reference the Rocket.Chat message:
direct-reply-method

**How it works**
When the `Subject` option is selected the subject of a reply email will look like
```
[Rocket.Chat local] You have been direct messaged by ruedigertest reply:N8K3QZyWcMCxZ4Gez
```

The email interceptor now check which option is activated and either use the email recipient or the subject to get the Rocket.Chat message id (see `/app/lib/server/lib/processDirectEmail.js` for details):
```
	if (settings.get('Direct_Reply_Method') && settings.get('Direct_Reply_Method') === 'subject') {
		// message id to reply to is encoded by email 'subject'
		const ind = email.headers.subject.indexOf('reply:');
		if (ind > 0) {
			email.headers.mid = email.headers.subject.substring(ind + 6, ind + 23);
		}
	} else if (settings.get('Direct_Reply_Method') && settings.get('Direct_Reply_Method') === 'to') {
		// message id to reply to is encoded by email 'to'
		// 'To' email format "username+messageId@domain"
		if (email.headers.to.indexOf('+') >= 0) {
			// Valid 'To' format
			email.headers.mid = email.headers.to.split('@')[0].split('+')[1];
		}
	} 
```

**Fixing the option `Delete Intercepted Emails`**
`interceptDirectReplyEmails.js` line 40:
```
this.delete = settings.get('Direct_Reply_Delete') ? settings.get('Direct_Reply_Delete') : true;
```
will never be false.

**Open questions/points**
- [ ] The filename `interceptDirectReplyEmails.js` does not fit the implementation details. Currently we have a email client implemented that fetches all new mails from the assigned email-address. Currently all new messages are marked as read, no matter if the e-mail is a "Rocket.Chat Mail" or a human message to the configured mailbox. Therefore the inbox should either be exclusively controlled by Rocket.Chat (no human access) or the implementation must be improved that only the minimum technical interaction takes place.

### CODE REVIEW

File: app/lib/server/lib/interceptDirectReplyEmails.js [hunk 1]
```diff
@@ -18,7 +37,7 @@ export class IMAPIntercepter {
 			keepalive: true,
 		});
 
-		this.delete = settings.get('Direct_Reply_Delete') ? settings.get('Direct_Reply_Delete') : true;
```
Comment (ruKurz) [inline app/lib/server/lib/interceptDirectReplyEmails.js]: will never be false

File: app/lib/server/lib/interceptDirectReplyEmails.js [hunk 2]
```diff
@@ -5,6 +5,25 @@ import POP3Lib from 'poplib';
...
+ *    - automatically send an livechat invitation via email to the customer
+ *
+ * 2) Synchronize an email with chatbased conversations
+ *    - Writing emails to an named mailbox will create new chat message in a specified channel
+ *    - Writing chat message within that channel will also send a mail to the according mailbox
+ *
+ * 3) Direct reply to offline messages via mail
+ *    - Sending a chat message while users are offline will cause an email notification
+ *    - Direct reply allows to answer to the chat message by reply to the email notification
+ */
```
Comment (ruKurz) [inline app/lib/server/lib/interceptDirectReplyEmails.js]: there are a lot more possibilities to use email in. Slack has just announced an email in/out:
https://slackhq.com/new-slack-features-boost-alignment-productivity

File: app/lib/server/lib/interceptDirectReplyEmails.js [hunk 3]
```diff
@@ -117,19 +136,32 @@ export class IMAPIntercepter {
...
 
 					// On fetched each message, pass it further
 					msg.once('end', Meteor.bindEnvironment(() => {
-						// delete message from inbox
-						if (this.delete) {
-							this.imap.seq.addFlags(seqno, 'Deleted', (err) => {
-								if (err) { console.log(`Mark deleted error: ${ err }`); }
-							});
+						// direct reply emails must have a message id (in its 'to' or 'subject' header) and a need a reference (in-reply-to)
+						if ((email.headers.subject.indexOf('reply:') > 0 || email.headers.to.indexOf('+') > 0) && email.headers.references) {
```
Comment (ruKurz) [inline app/lib/server/lib/interceptDirectReplyEmails.js]: only handle those emails having a reply messages referenced and a a reference (in-reply-to). Leave all other mails untouched by Rocket.Chat.

File: app/lib/server/lib/interceptDirectReplyEmails.js [hunk 4]
```diff
@@ -117,19 +136,32 @@ export class IMAPIntercepter {
...
+							// On fetched each message, pass it further
+							if (this.delete) {
+								this.imap.seq.addFlags(seqno, 'Deleted', (err) => {
+									if (err) { console.log(`Mark deleted error: ${ err }`); }
+								});
+							} else {
+								this.imap.seq.addFlags(seqno, 'Flagged', (err) => {
```
Comment (ruKurz) [inline app/lib/server/lib/interceptDirectReplyEmails.js]: If the handled mails are not being deleted: Mark them, in order to make it easier for humans to check the e-mail inbox.

File: app/lib/server/lib/interceptDirectReplyEmails.js [hunk 5]
```diff
@@ -117,19 +136,32 @@ export class IMAPIntercepter {
...
+									if (err) { console.log(`Mark as flagged error: ${ err }`); }
+								});
+							}
+							processDirectEmail(email);
```
Comment (ruKurz) [inline app/lib/server/lib/interceptDirectReplyEmails.js]: What is happening if the processing goes wrong? Errors should be handled better!

**Comment by ruKurz** [comment]
@Hudell 

**What I don't like so much about this PR is that it mixes several things**
1. try to support a mixed mailbox that can be handled by people and machines
2. only messages referencing a message id are processed, which leaves a lot of unprocessed mails in the inbox if no one has access to the mailbox
3. places different actions in the room after processing the e-mails, but implements only one of the options (direct reply)
4. criticizes the basic semantics without correcting them.

I just wanted to resume my work on this PR or start a new "smaller" PR in order to better separate the above points from each other. Just in this moment I saw that you assigned this PR to yourself.

May I assume that you would like to finalize the addition of the direct reply function via e-mail subject yourself? Or can I help you with the preliminary work?

**Comment by Hudell** [comment]
I was planning on reviewing and assisting only.

Splitting the PR into smaller ones would make it easier to review and merge them one by one too, so I'm ok with that.

**Comment by ruKurz** [comment]
@Hudell Now I have started a new PR that contains only the most necessary changes to support direct reply via email subject:
https://github.com/RocketChat/Rocket.Chat/pull/14697

Therefore this PR is now superfluous.

**Pull request closed without merging**